File: src/035/p035.py
import itertools
from src.utils.py.prime import generate_primes, is_prime


def is_circle_prime(digits: tuple[int, ...], primes: list[int]) -> bool:
    digit_str = "".join(str(d) for d in digits)
    for _ in range(len(digits)):
        if not is_prime(int(digit_str), primes):
            return False
        digit_str = digit_str[1:] + digit_str[0]
    return True


def solve() -> int:
    primes = generate_primes(1000)

    # Count single-digit primes separately
    count = len(list(itertools.takewhile(lambda n: n < 10, primes)))

    for length in range(2, 7):
        for combo in itertools.product([1, 3, 7, 9], repeat=length):
            if is_circle_prime(combo, primes):
                count += 1
    return count

File: src/P035/p035.go
package problem035

import (
	"strconv"

	"github.com/sunilbpandey/project-euler/src/utils/go/math/primeutils"
	"github.com/sunilbpandey/project-euler/src/utils/go/sliceutils"
	"github.com/sunilbpandey/project-euler/src/utils/go/strutils"
)

func isCircularPrime(digits []int, primes []int) bool {
	str := strutils.JoinInts(digits)
	for i := 0; i < len(digits); i++ {
		if !primeutils.IsPrime(strutils.Atoi(str), primes) {
			return false
		}
		str = str[1:] + str[:1]
	}
	return true
}

func Solve() string {
	primes := primeutils.GeneratePrimes(1000)

	// We are given that 13 circular primes exist below 100
	count := 13

	for length := 3; length < 7; length++ {
		for digits := range sliceutils.Product([]int{1, 3, 7, 9}, length) {
			if isCircularPrime(digits, primes) {
				count++
			}
		}
	}
	return strconv.Itoa(count)
}
